sound/soc/sof/intel/hda-dai.c

```diff
@@ -392,16 +408,40 @@ static int hda_dai_hw_free(struct snd_pcm_substream *substream,
 		snd_hdac_ext_link_clear_stream_id(link, stream_tag);
 	}
 
-	snd_soc_dai_set_dma_data(dai, substream, NULL);
+	snd_soc_dai_set_dma_data(cpu_dai, substream, NULL);
 	snd_hdac_ext_stream_release(link_dev, HDAC_EXT_STREAM_TYPE_LINK);
 	link_dev->link_prepared = 0;
 
 	/* free the host DMA channel reserved by hostless streams */
+	hda_stream = hstream_to_sof_hda_stream(link_dev);
 	hda_stream->host_reserved = 0;
 
 	return 0;
 }
 
+static int hda_dai_hw_free(struct snd_pcm_substream *substream,
+			   struct snd_soc_dai *dai)
+{
+	struct snd_soc_dapm_widget *w;
+	int ret;
+
+	ret = hda_dai_link_hw_free(substream);
```

Comment:
The order how the free was done also changes, before the `hda_dai_widget_update(dai->dev, w, DMA_CHAN_INVALID, false);` was done _before_ what it is now under `hda_dai_link_hw_free()`
Is there any dependency between the two?
For me the operation done under `hda_dai_link_hw_free()` is more like platform driver like (DMA free) which might come from the way HDA is architectured..

Reply:
I don't think there's any relationship between dealing with the widget and dealing with the stream tag, but that's precisely what I wanted to poke at.